File: Java/프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 0;
        
        for (int i = 0; i < s.length(); i++) {
            String rotated = rotateLeft(s, i);
            if (isCorrect(rotated.toCharArray())) {
                answer += 1;
            }
        }
        
        return answer;
    }
    
    public String rotateLeft(String s, int n) {
        return s.substring(n) + s.substring(0, n);
    }
    
    // 올바른 괄호 유무 판단 함수
    public boolean isCorrect (char[] str) {
        ArrayDeque<Character> stack = new ArrayDeque<>();
        
        for (char a : str) {
            if (a == '(' || a =='{' || a == '[') {
                stack.push(a);
            } else {
                if (a == ')') {
                    if (stack.isEmpty() || stack.pop() != '(') {
                        return false;
                    }
                } else if (a == '}') {
                    if (stack.isEmpty() || stack.pop() != '{') {
                        return false;
                    }
                } else if (a == ']') {
                    if (stack.isEmpty() || stack.pop() != '[') {
                        return false;
                    }
                }
            }
        }
        return stack.isEmpty();
    }
}

File: JavaScript/프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.js
function solution(s) {
    let answer = 0;
    for (let i = 0; i < s.length; i++) {
        let newStr = s.substring(i) + s.substring(0, i);
        let stack = [];
        let isCorrect = true;
        for (let c of newStr) {
            if (c == "(" || c == "{" || c == "[") {
                stack.push(c);
            } else if (c == ")") {
                if (stack.length == 0 || stack.pop() != "(") {
                    isCorrect = false;
                    break;
                }
            } else if (c == "}") {
                if (stack.length == 0 || stack.pop() != "{") {
                    isCorrect = false;
                    break;
                }
            } else if (c == "]") {
                if (stack.length == 0 || stack.pop() != "[") {
                    isCorrect = false;
                    break;
                }
            }
        }
        if (isCorrect && stack.length === 0) {
            answer += 1;
        }
    }
    return answer;
}